wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/JpaWorkspaceActivityDao.java

```diff
@@ -38,28 +36,170 @@
 
   @Override
   public void setExpiration(WorkspaceExpiration expiration) throws ServerException {
-    requireNonNull(expiration, "Required non-null expiration object");
+    setExpirationTime(expiration.getWorkspaceId(), expiration.getExpiration());
+  }
+
+  @Override
+  public void setExpirationTime(String workspaceId, long expirationTime) throws ServerException {
+    requireNonNull(workspaceId, "Required non-null workspace id");
+    doUpdate(workspaceId, a -> a.setExpiration(expirationTime));
+  }
+
+  @Override
+  public void removeExpiration(String workspaceId) throws ServerException {
+    requireNonNull(workspaceId, "Required non-null workspace id");
+    doUpdateOptionally(workspaceId, a -> a.setExpiration(null));
+  }
+
+  @Override
+  public List<String> findExpired(long timestamp) throws ServerException {
     try {
-      doCreateOrUpdate(expiration);
+      return doFindExpired(timestamp);
     } catch (RuntimeException x) {
       throw new ServerException(x.getLocalizedMessage(), x);
     }
   }
 
   @Override
-  public void removeExpiration(String workspaceId) throws ServerException {
-    requireNonNull(workspaceId, "Required non-null id");
+  @Transactional(rollbackOn = ServerException.class)
+  public void removeActivity(String workspaceId) throws ServerException {
+    EntityManager em = managerProvider.get();
```

Comment:
I assume manager provider may throw a RuntimeException. Please consider moving it inside of try/catch statement